test/shape_attributes.cc

```diff
@@ -0,0 +1,152 @@
+        {"lat":52.09050,"lon":5.09769},
+        {"lat":52.09098,"lon":5.09679}
+      ],"costing":"auto","shape_match":"map_snap"})");
+
+  rapidjson::Document doc;
+  doc.Parse(result_json);
+  if (doc.HasParseError()) {
+    throw std::logic_error("");
+  }
+
+  auto shape =
+      midgard::decode<std::vector<PointLL>>(rapidjson::Pointer("/shape").Get(doc)->GetString());
+  auto shape_attributes_time = rapidjson::Pointer("/shape_attributes/time").Get(doc)->GetArray();
+  auto shape_attributes_length = rapidjson::Pointer("/shape_attributes/length").Get(doc)->GetArray();
+  auto shape_attributes_speed = rapidjson::Pointer("/shape_attributes/speed").Get(doc)->GetArray();
+  auto edges = rapidjson::Pointer("/edges").Get(doc)->GetArray();
+
+  if (shape_attributes_time.Size() != shape.size() - 1)
+    throw std::logic_error("Expected: " + std::to_string(shape.size() - 1) +
+                           " | Found: " + std::to_string(shape_attributes_time.Size()));
+  if (shape_attributes_length.Size() != shape.size() - 1)
+    throw std::logic_error("Expected: " + std::to_string(shape.size() - 1) +
+                           " | Found: " + std::to_string(shape_attributes_length.Size()));
+  if (shape_attributes_speed.Size() != shape.size() - 1)
+    throw std::logic_error("Expected: " + std::to_string(shape.size() - 1) +
+                           " | Found: " + std::to_string(shape_attributes_speed.Size()));
+
+  // Measures the length between point
+  for (int i = 1; i < shape.size(); i++) {
+    auto distance = shape[i].Distance(shape[i - 1]) * .001f;
+
+    // Measuring that the length between shape pts is approx. to the shape attributes length
+    if (!midgard::equal(distance, shape_attributes_length[i - 1].GetFloat(), .01f)) {
+      throw std::logic_error(
+          "Expected: " + std::to_string(shape_attributes_length[i - 1].GetFloat()) +
+          " | Found: " + std::to_string(shape[i].Distance(shape[i - 1]) * .001f));
+    }
+  }
+
+  // Assert that the shape attributes (time, length, speed) are equal to their corresponding edge
+  // attributes
+  for (int e = 0; e < edges.Size(); e++) {
+    auto edge_length = edges[e]["length"].GetDouble();
+    auto edge_speed = edges[e]["speed"].GetDouble();
+
+    double sum_times = 0;
+    double sum_lengths = 0;
+    for (int j = edges[e]["begin_shape_index"].GetInt(); j < edges[e]["end_shape_index"].GetInt();
+         j++) {
+      sum_times += shape_attributes_time[j].GetDouble();
+      sum_lengths += shape_attributes_length[j].GetDouble();
+
+      if (!midgard::equal(edge_speed, shape_attributes_speed[j].GetDouble(), .15)) {
+        throw std::logic_error("Expected shape_attributes.speed to approx equal " +
+                               std::to_string(shape_attributes_speed[j].GetFloat()) +
+                               " | Found: " + std::to_string(edge_speed));
+      }
+    }
+
+    // Can't assert that sum of shape times equals edge's elapsed_time because elapsed_time includes
```

Comment:
it would be cool to add a test case where we can be sure turn costs wont impact. pick a long edge and put a point at the beginning and end. it wont go t hrough any intersections so you wont incur any turn costs. this also lets us test the single edge special case.